src/change_stream.ts

```diff
@@ -655,11 +629,22 @@ export class ChangeStream<
   hasNext(callback: Callback<boolean>): void;
   hasNext(callback?: Callback): Promise<boolean> | void {
     this._setIsIterator();
-    return maybePromise(callback, cb => {
-      this._getCursor((err, cursor) => {
-        if (err || !cursor) return cb(err); // failed to resume, raise an error
-        cursor.hasNext(cb);
-      });
+    return maybePromise(callback, callback => {
+      (async () => {
+        try {
+          const hasNext = await this.cursor.hasNext();
+          return callback(undefined, hasNext);
+        } catch (error) {
+          const errorOnResume = await this._processErrorAsync(error).catch(err => err);
+          if (errorOnResume) return callback(errorOnResume);
+          try {
+            const hasNext = await this.cursor.hasNext();
```

Comment:
am I reading it correctly here that in the case of a resumable error, we attempt to resume exactly once? if so, is that explicit spec behavior? (same question on the other methods here)

Reply:
Yes, this is spec behavior - https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#why-are-changestreams-required-to-retry-once-on-a-resumable-error

The idea is that some errors might be transient, but might also indicate an actual problem.  So we attempt to resume once in case it's not a problem, but if it fails a second time we propagate the error to the user.